File: [0089][Medium][Gray_Code]/Gray_Code_2.java
// leetcode time     cost : 1 ms
// leetcode memory   cost : 37.9 MB
// solution 2 for space saving,
class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> dp = new ArrayList<>(1 << n);
        dp.add(0);
        for (int num = 1; num <= n; num++) {
            int count = 1 << num;
            int toAdd = 1 << (num - 1);
            for (int i = count / 2; i < count; i++) {
                int value = dp.get(count - 1 - i) + toAdd;
                dp.add(value);
            }
        }
        return dp;
    }
}

public class Gray_Code_2 {
    public static void main(String args[]) {
        int num = 2;                     // #expect is [0,1,3,2]
        Solution Solution_obj = new Solution();
        Boolean result = Solution_obj.grayCode(num);
        System.out.println("return value is :" + (result));
    }
}

File: [0089][Medium][Gray_Code]/Gray_Code_2.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 32 ms
# leetcode memory   cost : 13.7 MB
# solution 1, 
# solution 2, bit operation
class Solution:
    def grayCode(self, n: int) -> [int]:
        res = [0]
        for i in range(n):
            for j in range(len(res) - 1, -1, -1):
                res.append(res[j] ^ (1 << i))
        return res
    
def main():
    code = 2             #expect is [0,1,3,2]
    obj = Solution()
    result = obj.grayCode(code)
    print("return result is :",result)
    
if __name__ =='__main__':
    main()   